kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -757,87 +760,327 @@ object Resource extends ResourceInstances with ResourcePlatform {
-      implicit F0: MonadCancel[F, Throwable],
-      K0: SemigroupK[F],
-      G0: Ref.Make[F]): ResourceSemigroupK[F] =
-    new ResourceSemigroupK[F] {
-      def F = F0
-      def K = K0
-      def G = G0
+// TODO the rest of the instances
+abstract private[effect] class ResourceMonadCancel[F[_]]
+    extends ResourceMonadError[F, Throwable]
+    with MonadCancel[Resource[F, *], Throwable] {
+  implicit protected def F: MonadCancel[F, Throwable]
+
+  def canceled: Resource[F, Unit] =
+    Resource.eval(F.canceled)
+
+  def forceR[A, B](fa: Resource[F, A])(fb: Resource[F, B]): Resource[F, B] =
+    Resource.applyFull(
+      _(fa.use_ !> fb.allocated).map(_.map(fin => (_: Resource.ExitCase) => fin)))
+
+  /*
+   * The problem is the scoping. I really need to be able to wrap a scope around
+   * a given Resource which is not extended by flatMap. This would be easy except
+   * for the fact that the existing resource itself defines scopes which
+   * themselves must be extended by flatMap, despite the closure of the onCancel
+   * scope.
+   */
+  def onCancel[A](fa: Resource[F, A], fin: Resource[F, Unit]): Resource[F, A] =
+    Resource applyFull { poll =>
+      poll(fa.allocated).onCancel(fin.use_).map(_.map(fin => (_: Resource.ExitCase) => fin))
+    }
+
+  def uncancelable[A](body: Poll[Resource[F, *]] => Resource[F, A]): Resource[F, A] =
+    Resource applyFull { poll =>
+      val inner = new Poll[Resource[F, *]] {
+        def apply[B](rfb: Resource[F, B]): Resource[F, B] =
+          Resource.applyFull(
+            _(poll(rfb.allocated)).map(_.map(fin => (_: Resource.ExitCase) => fin)))
+      }
+
+      body(inner).allocated.map(_.map(fin => (_: Resource.ExitCase) => fin))
+    }
+}
+
+// note: Spawn alone isn't possible since we need Concurrent to implement start
+abstract private[effect] class ResourceConcurrent[F[_]]
+    extends ResourceMonadCancel[F]
+    with GenConcurrent[Resource[F, *], Throwable] {
+  implicit protected def F: Concurrent[F]
+
+  def never[A]: Resource[F, A] =
+    Resource.eval(F.never[A])
+
+  def cede: Resource[F, Unit] =
+    Resource.eval(F.cede)
+
+  /*
+   * 1. If the scope containing the *fiber* terminates:
+   *   a) if the fiber is incomplete, run its inner finalizers when it complets
+   *   b) if the fiber is succeeded, run its finalizers
+   * 2. If the fiber is canceled or errored, finalize
+   * 3. If the fiber succeeds, await joining and extend finalizers into joining scope
+   * 4. If a fiber is canceled *after* it completes, we explicitly snag the finalizers and run them, setting the result to Canceled()
+   *   a) Note that we are missing a check here! We need to disable this behavior if someone has already joined
+   */
+  def start[A](fa: Resource[F, A]): Resource[F, Fiber[Resource[F, *], Throwable, A]] = {
+    final case class State(
+        fin: F[Unit] = F.unit,
+        runOnComplete: Boolean = false,
+        completed: Boolean = false,
+        canceled: Boolean = false)
```

Comment:
```suggestion
    final case class State(
        fin: Option[F[Unit]] = None,
        runOnComplete: Boolean = false,
        canceled: Boolean = false,
        joined: Boolean = false)
```
If we use `allocated` below, we could represent the need to finalize as the presence of a finalizer. When we actually decide to finalize, we set `fin = None` allowing us to simplify the outer resource finalizer to just "run it if it's there".